### ISSUE TITLE
GraphQL support

### ISSUE THREAD

**Issue by njsmith** [brettcannon/gidgethub#89]
Github's GraphQL "v4" API seems to have some attractive properties, like being able to do complex custom queries in a single request. Also, Github support just told me that there's functionality that's *only* accessible through the GraphQL API. Specifically I asked how you use the API to add a reaction to a pull request review, which is [mysteriously missing from their docs](https://developer.github.com/v3/reactions/), and they told me:

> 
> Thanks for following up, Nathaniel. This is not possible via the REST API, but is possible via the GraphQL API using this mutation:
> 
> https://developer.github.com/v4/mutation/addreaction/
> 
> Notice the supported subject types:
> 
> https://developer.github.com/v4/interface/reactable/
> 
> Hope this helps.

This isn't super critical functionality or anything, but it does seem like we might want the option to use GraphQL more.

Has anyone already figured out how to use GraphQL in a gidgethub-based framework? I would love to see some examples, b/c I find the Github docs largely impenetrable :-).

Are there new features that gidgethub should grow to support GraphQL better? I have no idea but maybe someone does.

CC: @wgwz @webknjaz

**Comment by brettcannon** [comment]
My plan is to look at pre-existing GraphQL libraries to see if any of them work reasonably well before making a decision about what gidgethub may want to do in this area (which could vary from "do nothing" to "code generation for GraphQL queries that is fully typed"). Basically I'm thinking about a workflow where you figured out your queries in some playground, saved it to a `.gql` file, and then we auto-generate the Python code to make the request with complete type annotations for both the call and return value.

Otherwise @warsaw opened https://github.com/brettcannon/gidgethub/pull/84 (and I closed) if you need some code to use now.

**Comment by webknjaz** [comment]
@brettcannon 
Not sure if it makes sense to stick validation into an ever-changing GitHub's API structure. I like that gidgethub is minimalistic.

@njsmith have you tried just passing `gh_api.post(URL, data=<GQL_FORMATED_DATA_STR_HERE>)`? Since it doesn't only accept dict but also str/bytes AFAIR.

**Comment by njsmith** [comment]
@webknjaz Yeah judging from Barry's PR it looks like it's just a matter of calling something like

```python
gh.post("https://api.github.com/graphql", data={"query": GQL_QUERY_STRING, ...}, **authorization_args)
```

You don't actually pass the graphql query as the body directly; it gets packaged up as an attribute in a JSON document. I think. You might also be able to set other attributes inside that JSON object, like `variables: ...`? That's what the GraphQL Explorer does. The GH docs OTOH say that you should jam a variables section into the query string, which seems more awkward to me.

**Comment by brettcannon** [comment]
@webknjaz do understand that the solution I'm considering would rely heavily on GraphQL's queryable schema support with code generation, so updating your code if something changed would not be difficult (else I'm not interested). Plus any changes that would be hard would bump up against your query requiring changes anyway.

**Comment by brettcannon** [comment]
OK, so none of the Python GraphQL libraries do what I want:
1. Parse a GraphQL query
2. Take in a downloaded schema file
3. Generate Python code that is fully typed and emits a function that takes the appropriate arguments for a query with variables

As such, it probably makes sense to add a simple solution to gidgethub. Let me think about it a bit and once I have a design in mind I will come back here and make sure it covers everyone's needs (my initial guess is adding `def graphql(self, query: str, /, **variables: Any) -> Any` to `gidgethub.abc.GitHubAPI` will cover it).

**Comment by webknjaz** [comment]
The only thing that's bothering me a bit is that there's also schemas for REST (v3) which are harder to construct but I've seen it done in several projects.
It seems like adding this code generation *just* for GraphQL creates some sort of inconsistency from the UX perspective. Maybe this could be generalized/extended to REST too?

**Comment by brettcannon** [comment]
@webknjaz one key thing to know is any auto-generation is for GraphQL in _general_, while the REST stuff here is _specific_ to GitHub. IOW it would probably be a separate project from gidgethub if I'm ever motivated enough to actually code up my "perfect" solution to GraphQL client usage.

And note that I am unaware of GitHub publishing their REST schema everywhere. They have always designed it so that you use the URL templates that you get in webhooks and responses specifically so you didn't need to query or track any schema ahead of time.

**Comment by webknjaz** [comment]
> And note that I am unaware of GitHub publishing their REST schema everywhere

So this one is a bit tricky.
TL;DR There's a few projects under octokit (that's used by Probot) that convert GitHub docs (HTML) into JSON: https://github.com/octokit/webhooks, https://github.com/octokit/routes, https://github.com/octokit/graphql-schema.

I think it may be possible to reuse these but it's probably out of scope for gidgethub. FTR I collect all related tricks here https://github.com/sanitizers/octomachinery/issues/7#issuecomment-557315422 because I'm researching whether it's possible to automatically inject preview headers into corresponding API queries...

**Comment by brettcannon** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Add simple GraphQL support to gidgethub](https://trello.com/c/XoFN298l/283-add-simple-graphql-support-to-gidgethub)

**Comment by brettcannon** [comment]
My current thought on the API.

```python
class _GraphQLErrorType(TypedDict):

    errors: List[Dict[str, Any]]


class GraphQLError(GitHubException):

    """Error(s) resulting from calling the GraphQL v4 API."""

    def __init__(self, details: _GraphQLErrorType) -> None:
        """Extract the 'message' from the first error while saving all details."""
        message = details["errors"][0]["message"]
        super().__init__(message)
        self.details = details


def graphql(self, query: str, /, **variables: Any) -> Any:
    """Query the GraphQL v4 API.

    The `variables` argument collects any GraphQL variables to use in the query.

    Any result which contains an "errors" key will trigger the raising of a
    GraphQLError exception.
    """
```
